File: utils/auth_utils.py
from fastapi import Depends, HTTPException, status
from fastapi.security import OAuth2PasswordBearer
from jose import JWTError, jwt, ExpiredSignatureError
from models.models import User
from schemas.auth import TokenData
import os
import logging

# Logger setup
logging.basicConfig(level=logging.INFO)
logger = logging.getLogger(__name__)

# JWT Configuration
SECRET_KEY = os.getenv("SECRET_KEY", "your_secret_key")
ALGORITHM = os.getenv("ALGORITHM", "HS256")

oauth2_scheme = OAuth2PasswordBearer(tokenUrl="/api/v3/login")

async def get_current_user(token: str = Depends(oauth2_scheme)):
 
    credentials_exception = HTTPException(
        status_code=status.HTTP_401_UNAUTHORIZED,
        detail="Could not validate credentials",
        headers={"WWW-Authenticate": "Bearer"}
    )
    try:
        payload = jwt.decode(token, SECRET_KEY, algorithms=[ALGORITHM], audience="fastapi-users")
        username: str = payload.get("sub")
        if not username or not isinstance(username, str):
            logger.error("Invalid or missing 'sub' claim in token")
            raise credentials_exception
        token_data = TokenData(username=username)
    except ExpiredSignatureError as e:
        logger.error(f"Expired token error: {e}")
        raise HTTPException(
            status_code=status.HTTP_401_UNAUTHORIZED,
            detail="Token has expired",
            headers={"WWW-Authenticate": "Bearer"}
        )
    except JWTError as e:
        logger.error(f"JWT error: {e}")
        raise credentials_exception

    # Retrieve the user from the database
    try:
        user = User.objects(username=token_data.username).first()
        if user is None:
            logger.warning(f"User not found for username: {token_data.username}")
            raise credentials_exception
        logger.info(f"Authenticated user: {user.id}")
    except Exception as e:
        logger.error(f"Error authenticating user: {str(e)}")
        raise HTTPException(status_code=401, detail="Invalid authentication credentials")

    logger.info(f"User {token_data.username} successfully authenticated")
    return user

File: utils/auth_utils.js
const jwt = require('jsonwebtoken');
const { User } = require('../models/models');  // Assuming User is your Mongoose model
const { tokenDataSchema } = require('../schemas/auth');  // Assuming TokenData schema
const dotenv = require('dotenv');
const { Unauthorized } = require('http-errors');
const logger = require('pino')();  // Logger setup

dotenv.config();

// JWT Configuration
const SECRET_KEY = process.env.SECRET_KEY || 'your_secret_key';
const ALGORITHM = process.env.ALGORITHM || 'HS256';
const ACCESS_TOKEN_EXPIRE_MINUTES = parseInt(process.env.ACCESS_TOKEN_EXPIRE_MINUTES) || 30;

const createAccessToken = (data) => {
  const toEncode = { ...data };
  const expire = Math.floor(Date.now() / 1000) + ACCESS_TOKEN_EXPIRE_MINUTES * 60;
  toEncode.exp = expire;
  const encodedJwt = jwt.sign(toEncode, SECRET_KEY, { algorithm: ALGORITHM });
  return encodedJwt;
};

const oauth2Scheme = 'Bearer';  // For token validation

// Function to get current user based on JWT token
async function getCurrentUser(req, res, next) {
  const token = req.headers.authorization && req.headers.authorization.startsWith(oauth2Scheme)
    ? req.headers.authorization.split(' ')[1]
    : null;

  if (!token) {
    return next(new Unauthorized('Could not validate credentials'));
  }

  try {
    const payload = jwt.verify(token, SECRET_KEY, { algorithms: [ALGORITHM], audience: 'fastapi-users' });
    const username = payload.sub;

    if (!username || typeof username !== 'string') {
      logger.error("Invalid or missing 'sub' claim in token");
      throw new Unauthorized('Could not validate credentials');
    }

    const tokenDataSchema = new tokenDataSchema({ id });

    // Retrieve the user from the database
    const user = await User.findOne({ id: tokenDataSchema.id });

    if (!user) {
      logger.warn(`User not found for username: ${tokenDataSchema.id}`);
      throw new Unauthorized('Could not validate credentials');
    }

    logger.info(`Authenticated user: ${user._id}`);
    req.user = user;  // Attach user to request object

    next();
  } catch (err) {
    if (err instanceof jwt.TokenExpiredError) {
      logger.error(`Expired token error: ${err}`);
      return next(new Unauthorized('Token has expired'));
    }
    
    if (err instanceof jwt.JsonWebTokenError) {
      logger.error(`JWT error: ${err}`);
      return next(new Unauthorized('Could not validate credentials'));
    }

    logger.error(`Error authenticating user: ${err.message}`);
    return next(new Unauthorized('Invalid authentication credentials'));
  }
}

module.exports = { getCurrentUser, createAccessToken };
